Rework application, create static middleware

saralein opened the pull request.

saralein committed:
Rework application, create static middleware

zdavy commented on server/src/main/java/com/saralein/server/middleware/StaticMiddleware.java:
```diff
@@ -0,0 +1,81 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
```
This class feels like a router implementation that also handles requests for static resources rather than middleware.

zdavy replied:
I don't feel one way or another about using it that way, I just think the naming is a little misleading.

zdavy commented on server/src/main/java/com/saralein/server/middleware/StaticMiddleware.java:
```diff
@@ -0,0 +1,81 @@
+    public StaticMiddleware(Path root, Router router) {
+        this(new FileHelper(root), router, new DirectoryController(new FileHelper(root)),
+                new FileController(new FileHelper(root)));
+    }
```
Seems like all the file related stuff falls in here, it feels like partial content belongs in here too.

saralein replied:
Yeah, it will go here.  It doesn't exist on this branch yet though.  Perhaps I should have waited for it to open this.

zdavy replied:
Yeah good point, I guess once one gets merged the second'll get rebased.

saralein committed:
Add new interfaces for router and middleware

- Add Callable interface with call method for router/middleware
- Add Middleware interface with apply method for middleware

saralein committed:
Restore previously deleted error controller

damonkelley commented on server/src/main/java/com/saralein/server/Application.java:
```diff
@@ -1,48 +1,50 @@
-        ErrorController errorController = new ErrorController();
+        public Application build() {
+            Callable router = configureRouter();
+            Middleware staticMiddleware = new StaticMiddleware(root);
```
I would like to see this not hard-coded in the Application Builder. I feel like this should be treated more as configuration, and wired up inside of main. We don't have any other middleware right now, but the point of middleware, in a way, is to be open/closed.

damonkelley commented on server/src/main/java/com/saralein/server/Application.java:
```diff
@@ -1,48 +1,50 @@
+        private Router configureRouter() {
+            return this.routes == null
+                    ? new Router(new Routes())
+                    : new Router(routes);
```
Do we need this? Is there ever a time when `this.routes` is null? 

It would be great if we could not new up the router in the the Application. Aside from the Builder, it seems like the Application class doesn't need to have a dependency on Router, right? It could just depend on receiving a `Callable`?

damonkelley replied:
It looks like this is here for the tests?

damonkelley commented on server/src/main/java/com/saralein/server/router/Router.java:
```diff
@@ -2,74 +2,40 @@
-        Path resource = Paths.get(resourceUri);
-        resourceExists = Files.exists(resource);
-        resourceIsDirectory = Files.isDirectory(resource);
+    public Router(Routes routes) {
```
👏

damonkelley commented on server/src/main/java/com/saralein/server/Application.java:
```diff
@@ -1,48 +1,50 @@
-    public void start(int port, Path root) {
-        Runtime runtime = Runtime.getRuntime();
-        FileHelper fileHelper = new FileHelper(root);
+    public static class Builder {
```
I am curious, what does this Builder do for us? Does it make something easy? hard?

damonkelley reviewed (changes requested):
I like where this is going. Just a couple comments.

saralein committed:
Remove application builder

zdavy commented on server/src/main/java/com/saralein/server/router/Router.java:
```diff
@@ -2,74 +2,40 @@
-        }
+    public Router(Routes routes, ErrorController errorController) {
+        this.routes = routes;
+        this.errorController = errorController;
```
So it seems a little off to be using a controller outside of routing, maybe we can abstract error handling into another piece of middleware?

saralein replied:
Yes, I started addressing this on the next PR.

saralein committed:
Convert file/directory controllers to handlers

saralein committed:
Add final callable, adjust errors in middleware

saralein committed:
Incorporate error response in middleware

saralein committed:
Remove final callable
